### PR TITLE
Add CLI program option for Program ID

### PR DESCRIPTION
**Pull request by ngundotra**
Provide -p to override program address.

Also adds support for COptions by treating them as normal Option types. Closes https://github.com/metaplex-foundation/shank/issues/22.

Works with spl-token program via `cargo +1.59.0 run -- idl -r <YOUR-SYS-PATH>/solana-program-library/token/program -o ./tmp/token/idl -p TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA.`

### CODE REVIEW

File: shank-macro-impl/src/types/resolve_rust_ty.rs
Comment (thlorenz) [inline shank-macro-impl/src/types/resolve_rust_ty.rs:2]: I'm not sure about this file, looks like all format changes?
Maybe your IDE isn't respecting the `rustfmt.toml` in the package folder?
I just ran `cargo fmt` inside that package and don't get any changes.

File: shank-idl/src/file.rs
Comment (thlorenz) [inline shank-idl/src/file.rs:44]:
```suggestion
            program_address_override: None,
```
and all related renames.

File: shank-macro-impl/src/macros/program_id.rs
Comment (thlorenz) [inline shank-macro-impl/src/macros/program_id.rs:43]: This makes no sense here as if shank found multiple `program_id!`s in the code then providing another one via a flag surely won't fix that problem :)

File: shank-macro-impl/src/macros/program_id.rs
Comment (thlorenz) [inline shank-macro-impl/src/macros/program_id.rs:50]:
```suggestion
            "Could not find a `declare_id(\"<program-id>\")` invocation in the program. If this is intentional provide a program address via the -p argument instead"
```

As a result most test assertions need to be adapted

**Review by thlorenz — CHANGES_REQUESTED** [review]
Thanks this is great!
I commented on the things I'd like to improve.

Also we should figure out what we do if the arg is provided _and_ we find a `program_id!` decl in the code.
Will it just override it? I think that'd make sense.

We should probably add this info to the CLI docs [here](https://github.com/metaplex-foundation/shank/blob/master/shank-cli/README.md)

File: shank-macro-impl/src/macros/program_id.rs
Comment (ngundotra) [reply shank-macro-impl/src/macros/program_id.rs:43]: I think it makes sense if the default behavior is to always override the program id.

If users want to be notified if they are going to accidentally override the program id, they should pass "--warn" or something

Comment (thlorenz) [reply shank-macro-impl/src/macros/program_id.rs:43]: OK, but then we'd need to indicate that clearer in the message:

```
Found more than one program id candidate: {:?}. You should either have exactly one `declare_id!` in your code or override the program id via -p.
```

File: shank-macro-impl/src/macros/program_id.rs
Comment (ngundotra) [reply shank-macro-impl/src/macros/program_id.rs]: Updated

File: shank-macro-impl/src/macros/program_id.rs
Comment (ngundotra) [reply shank-macro-impl/src/macros/program_id.rs]: Added this exact error message + updated tests

File: shank-macro-impl/src/types/resolve_rust_ty.rs
Comment (ngundotra) [reply shank-macro-impl/src/types/resolve_rust_ty.rs]: Ran `cargo fmt` and fixed the formatting issues. Sorry about that 😅

Change to support `COption` was that I just made it parse as a normal `Option` in `resolve_rust_ty.rs`
 ```rust
 "COption" | "Option" => {
 ```

File: shank-macro-impl/src/types/resolve_rust_ty.rs
Comment (thlorenz) [inline shank-macro-impl/src/types/resolve_rust_ty.rs:305]: This seems unrelated to the PR?

I agree we need to support different option types, but I wanted to first verify that `COption` is indeed serialized the same way as `Option`.
Please remove this for now .. we can address in a separate PR.

I overlooked this change previously due to all the format changes ...

**Review by thlorenz — CHANGES_REQUESTED** [review]
LGTM once we remove the unrelated `COption` change.
I really wanna make sure it's the same as `Option` first and wanna address that separately.

Once that part is removed it's good to merge, thanks!

File: shank-macro-impl/src/types/resolve_rust_ty.rs
Comment (ngundotra) [reply shank-macro-impl/src/types/resolve_rust_ty.rs]: Awesome, removed the COption

File: shank-macro-impl/src/types/resolve_rust_ty.rs
Comment (thlorenz) [reply shank-macro-impl/src/types/resolve_rust_ty.rs]: Yeah I do wanna support all option types as well, but need to make sure first that they actually are de/serialized the same by borsh.
Need to add a [test case here](https://github.com/metaplex-foundation/beet/blob/master/tools/bsamples/src/options.rs).

**Pull request merged**
